src/storage/src/render/upsert/rocksdb.rs

```diff
@@ -100,42 +77,23 @@ impl UpsertStateBackend for RocksDB {
         R: IntoIterator<Item = &'r mut UpsertValueAndSize>,
     {
         let mut g_stats = GetStats::default();
-        let mut gets = gets.into_iter().peekable();
-        if gets.peek().is_some() {
-            let gets = gets.chunks(BATCH_SIZE);
-            let results_out = results_out.into_iter().chunks(BATCH_SIZE);
-
-            for (gets, results_out) in gets.into_iter().zip_eq(results_out.into_iter()) {
-                self.multi_get_scratch.clear();
-                self.multi_get_result_scratch.clear();
-                self.multi_get_scratch.extend(gets);
-                self.multi_get_result_scratch
-                    .extend((0..self.multi_get_scratch.len()).map(|_| None));
-
-                let stats = self
-                    .rocksdb
-                    .multi_get(
-                        self.multi_get_scratch.drain(..),
-                        self.multi_get_result_scratch.iter_mut(),
-                    )
-                    .await?;
+        let stats = self
+            .rocksdb
+            .multi_get(gets, results_out, |place, value| {
+                *place = value.map_or(
```

Comment:
OOC why was it necessary? Is it some memory optimization that it's doing a map and setting the value in the place right away. Instead of doing say like a separate map here on the values?

Reply:
Also, then could the closure just take value? And instead of `placement(place, get)`, it could do `*place = fn(get)` here https://github.com/MaterializeInc/materialize/pull/19509/files#diff-8e949ab69256efd04acb76b94b8ee2668d4a14501e7cc3a6d0325572a39b268fL366

Reply:
It's because `UpsertState::multi_get` takes an iterator of `&mut UpsertValueAndSize` and `RocksDBInstance::multi_get` produces `GetResult`'s, so it needs to be told how to place a `GetResult` into a `&mut UpsertValueAndSize` location

Reply:
So basically taking placement of type `Placement: Fn(Option<GetResult<V>>) -> Ret` should work as well (maybe with a better name)?

Then assignment can still inside the rocksdb crate `*place = placement_or_something_else(get)` and will not be exposed here. This method here then can only pass what the mapping would be from value to something else without required a reference to place.

```
.multi_get(gets, results_out, |value| { value.map_or (...)|
```

Reply:
(And I realized mapping separately here without passing any closure would mean allocating a vector of `GetResult`s here, passing it to get it populated and, then do the type conversion. Which would be bad)

Reply:
@moulimukherjee in reality the extra vector wouldnt be that bad (because we could re-use is across different calls), but this is technically more efficient, and we have to manage slightly less state

Reply:
> Then assignment can still be inside the [rocksdb crate](https://github.com/MaterializeInc/materialize/pull/19509/files#diff-8e949ab69256efd04acb76b94b8ee2668d4a14501e7cc3a6d0325572a39b268fL366) *place = to_type(get) and will not be exposed here. This method here then can only pass what the mapping would be from value to something else without requiring a reference to place.

Ah I see your point here...this would probably work! I'll check monday!

Reply:
@moulimukherjee this ended up being cleaner, thanks!